src/main/kotlin/org/codefreak/codefreak/service/file/FileSystemFileService.kt

```diff
@@ -0,0 +1,438 @@
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty
+import org.springframework.stereotype.Service
+
+@Service
+@ConditionalOnProperty(name = ["codefreak.files.adapter"], havingValue = "FILE_SYSTEM")
+class FileSystemFileService(@Autowired val config: AppConfiguration) : FileService {
+
+  companion object {
+    val blacklistedPaths = setOf(
+      ".git",
+      ".gitignore",
+      ".gitattributes"
+    )
+  }
+
+  override fun readCollectionTar(collectionId: UUID): InputStream {
+    val output = ByteArrayOutputStream()
+    val tarOutput = TarArchiveOutputStream(output)
+    val collectionPath = getCollectionPath(collectionId)
+
+    Files.walkFileTree(collectionPath, object : SimpleFileVisitor<Path>() {
+      override fun preVisitDirectory(dir: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        dir?.let {
+          val path = collectionPath.relativize(dir).normalize()
+          val normalizedPath = TarUtil.normalizeDirectoryName(path.toString())
+          TarUtil.mkdir(normalizedPath, tarOutput)
+        }
+        return FileVisitResult.CONTINUE
+      }
+
+      override fun visitFile(file: Path?, attrs: BasicFileAttributes?): FileVisitResult {
+        file?.let {
+          val path = collectionPath.relativize(file).normalize()
+          val normalizedPath = TarUtil.normalizeFileName(path.toString())
+          TarUtil.writeFileWithContent(normalizedPath, Files.newInputStream(it), tarOutput)
+        }
+        return FileVisitResult.CONTINUE
+      }
+    })
+
+    tarOutput.close()
+
+    return ByteArrayInputStream(output.toByteArray())
+  }
+
+  override fun writeCollectionTar(collectionId: UUID): OutputStream {
+    return object : ByteArrayOutputStream() {
+      override fun close() {
+        // Overwrite the existing collection with the tar contents
+        deleteCollection(collectionId)
+
+        TarArchiveInputStream(ByteArrayInputStream(toByteArray())).use { input ->
+          input.entrySequence().forEach {
+            val name = it.name
+            require(isValidPath(collectionId, Paths.get(it.name, ""))) { "`$name` is not a valid path" }
+
+            if (it.isFile) {
+              if (it.size == 0L) {
+                val fileName = TarUtil.normalizeFileName(it.name)
+                // create parent directories first because they might only implicitly exist in the tar through this file name
+                createDirectories(collectionId, setOf(FileUtil.getParentDir(fileName)))
+                createFiles(collectionId, setOf(fileName))
+              } else {
+                writeFile(collectionId, it.name).use { outputStream ->
+                  IOUtils.copy(input, outputStream)
+                }
+              }
+            } else if (it.isDirectory) {
+              createDirectories(collectionId, setOf(it.name))
+            }
+          }
+        }
+      }
+    }
+  }
+
+  private fun isValidPath(collectionId: UUID, path: Path): Boolean {
```

Comment:
I'm not a fan of this checking for valid paths. The collection should act like a chroot environment, so paths like `/dir/../../../` should resolve to `/` (collection root) like on a real file system. IMO there is no need to bail out if the path looks invalid. Simply make it a valid path and try to perform the operation.